liteeth/mac/__init__.py

```diff
@@ -31,8 +33,157 @@ def __init__(self, phy, dw,
             self.comb += Port.connect(self.interface, self.core)
             self.ev, self.bus = self.interface.sram.ev, self.interface.bus
             self.csrs = self.interface.get_csrs() + self.core.get_csrs()
+        elif interface == "hybrid":
+            # Wishbone MAC
+            self.rx_slots = CSRConstant(nrxslots)
+            self.tx_slots = CSRConstant(ntxslots)
+            self.slot_size = CSRConstant(2**bits_for(eth_mtu))
+            self.submodules.interface = LiteEthMACWishboneInterface(cpu_dw, nrxslots, ntxslots, endianness)
+            # HW accelerated MAC
+            self.submodules.crossbar = LiteEthMACCrossbar(dw)
+            # MAC crossbar
+            self.submodules.mac_crossbar = LiteEthMACCoreCrossbar(self.core, self.crossbar, self.interface, dw, cpu_dw, endianness, hw_mac)
+            # Connections
+            self.ev, self.bus = self.interface.sram.ev, self.interface.bus
+            self.csrs = self.interface.get_csrs() + self.core.get_csrs()
         else:
             raise NotImplementedError
 
     def get_csrs(self):
         return self.csrs
+
+class LiteEthMACCoreCrossbar(Module):
+    def __init__(self, core, crossbar, interface, dw, cpu_dw, endianness, hw_mac=None):
+        fifo_depth = 2048
+        wishbone_rx_fifo = stream.SyncFIFO(eth_phy_description(dw), fifo_depth, buffered=True)
+        wishbone_tx_fifo = stream.SyncFIFO(eth_phy_description(dw), fifo_depth, buffered=True)
+        crossbar_rx_fifo = stream.SyncFIFO(eth_phy_description(dw), fifo_depth, buffered=True)
+        crossbar_tx_fifo = stream.SyncFIFO(eth_phy_description(dw), fifo_depth, buffered=True)
+
```

Comment:
@piotr-binkowski: i'm just having a closer look at the code, can you explain the purpose of these FIFOs? I just want to see if we can avoid them since they are relatively large and will use Block RAMs.

Reply:
I think that those FIFOs are just a leftover from my early attempts at making the crossbar and removing them shouldn't break anything. You might want to check if it still works litex_sim as I had some issues with Verilator when using hw_mac configuration that were solved by adding small FIFOs there

Reply:
Ok thanks, i'll look at that.